### ISSUE TITLE
Failing Job 'run lastz' - Running Toil on AWS

### ISSUE THREAD

**Issue by tinaveit** [DataBiosphere/toil#4773]
Hi,
We are performing genome alignments using Progressive Cactus software and toil on Amazon Web Services.We don't have much computational background, and are getting the following error (several times for different runs):

`No log file is present, despite job failing: 'run_lastz' 41ff109d-df4c-45eb-897b-a7c2c9eae970 v1 Due to failure we are reducing the remaining try count of job 'run_lastz' 41ff109d-df4c-45eb-897b-a7c2c9eae970 v1 with ID 41ff109d-df4c-45eb-897b-a7c2c9eae970 to 5 We have increased the default memory of the failed job 'run_lastz' 41ff109d-df4c-45eb-897b-a7c2c9eae970 v1 to 2147483648 bytes We have increased the disk of the failed job 'run_lastz' 41ff109d-df4c-45eb-897b-a7c2c9eae970 v1 to the default of 2147483648 bytes Issued job 'run_lastz' 41ff109d-df4c-45eb-897b-a7c2c9eae970 v2 with job batch system ID: 17361 and disk: 2.0 Gi, memory: 2.0 Gi, cores: 1, accelerators: [], preemptible: False Job failed with exit value 255: 'run_lastz' 257e6b07-2875-4f0f-b233-166bc7d02364 v1 Exit reason: 3`

Then it mentions that several jobs were lost. 

It doesn't seem to crash the Cactus run, but once workflow is 100%, there is no output file. We haven't deleted any files from the jobstore on AWS, and we are feeling lost on what this error could be related to. Moreover, I have the full logfile downloaded to my local in case it may be helpful as well. Any help/ideas would be greatly appreciated, especially cause each AWS run is costly. Thank you for your time!

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1488)
┆Issue Number: TOIL-1488

**Comment by adamnovak** [comment]
It looks like we should probably be logging that as `Exit reason: LOST` instead of `Exit reason: 3`: https://github.com/DataBiosphere/toil/blob/73f155efd71ace58df5e54b0bc6ccbe9e9422f25/src/toil/batchSystems/abstractBatchSystem.py#L53-L54

When jobs are "lost" that is supposed to mean that the machine that was supposed to be running them itself vanished. Usually on AWS this will be because it is a spot market machine, which Amazon is allowed to take back if someone outbids you for it.

It sounds like you set up a cluster with AWS spot instances, and the Cactus workflow says that the lastz jobs are OK to get preempted and can run on spot nodes. But, your spot nodes are not sticking around long enough to actually finish the lastz jobs; either the spot price is rising higher than your spot bid, or Amazon is taking them away for other reasons.

Maybe try raising your spot bid, or else not using spot instances and paying for all reserved instances?

**Comment by adamnovak** [comment]
When you say "workflow is 100%", do you mean that the progress bar Toil shows hit 100%, or that the Toil process reported that the workflow completed successfully and returned a 0 exit code? The progress bar is not actually very good; it only knows about the jobs that have already been created in the workflow and doesn't predict what jobs will be created by those jobs in the future. So for a lot of workflows it will reach 100% because all the jobs it knows about have run, and then go back down when it looks and sees new jobs have been added to do.

**Comment by glennhickey** [comment]
Following up from email from @tinaveit .  The command line is apparently 
```
cactus --consCores 2 --nodeTypes c4.8xlarge,r4.8xlarge --minNodes 0,0 --maxNodes 20,1 --nodeStorage 250 --provisioner aws --batchSystem mesos --metrics aws:us-east-1:xtremo-devo-cactus --logFile cactus.log seqFile.txt output.hal
```
This means that spot nodes *are not* being used (as the `--nodeTytpes` arguments don't have `:<bid>` suffixes).  Not sure what's happening, are you able to post the whole log?

**Comment by glennhickey** [comment]
This log looks fine.  It's normal for lastz jobs to get swallowed up like that (I don't know why) but they were successfully rescheduled.  At the end of your log, it looked like it was running along fine in the "bar" stage.   A few things that did look suspicious:
* you specified `--consCores 2`.  That means the cactus job (which is running at the end of your log) only gets 2 threads.  So something I'd expect to take about 2 hours for 700mb genomes on 64 cores may take 30x longer.  
* your tree `(Oryzias_latipes:0.121261,Nematolebias_whitei:0.141531)0.471387;` gives the root node name `0.471387`.  This is supposed to trigger an error message in Cactus.  It didn't so it'll probably run through, but you want to avoid this in general.  (note that distances come after `:` chracters but there's no node above it so your input tree should just be `(Oryzias_latipes:0.121261,Nematolebias_whitei:0.141531);`

tldr: it seems to be running fine -- if it seems too  slow give it more cores with `--consCores`.

**Comment by tinaveit** [comment]
Okay, sounds good. Thanks so much for your input. I will change the tree and add some threads. Any suggestions of what would be an adequate number of --consCores? We had to add that argument, cause it was giving us errors in the beginning

**Comment by glennhickey** [comment]
`--consCores` should be equal to the number of cores on the AWS instance you are running on.  I guess 32 here?

**Comment by tinaveit** [comment]
@glennhickey I ran today with the updated consCores value and changed the tree. The cactus run crashed due to a failed job, and exited the instance. I have attached the end of the log file, where it shows the error and the failures. Everything above that looked normal. Not sure what could be going wrong now. Thanks for your time!
[cactus_end.log](https://github.com/DataBiosphere/toil/files/14560783/cactus_end.log)

**Comment by glennhickey** [comment]
This one looks like a bug
```
            return max(min(int(os.environ['CACTUS_MAX_MEMORY']), memory_bytes), int(os.environ['CACTUS_DEFAULT_MEMORY']))
        ValueError: invalid literal for int() with base 10: 'None'
```
I added these environment variables in the most recent version of Cactus (they are set by Cactus).  They work fine locally and on slurm, but if I'm reading this right are somehow getting lost on AWS (which is not part of our automated tests).  

Thanks for raising this -- I will try to reproduce it here and find a solution.

**Comment by adamnovak** [comment]
We're meant to snapshot the environment when the workflow starts up, and then ship that off to the worker to restore when running jobs:

https://github.com/DataBiosphere/toil/blob/0446fe285fe57a3d920d4b5a30d00f2a5aa8589c/src/toil/options/common.py#L680-L682

We capture `os.environ` inside `start()` here, and save to `environment.pickle` in the job store:

https://github.com/DataBiosphere/toil/blob/0446fe285fe57a3d920d4b5a30d00f2a5aa8589c/src/toil/common.py#L898

And we restore it here in the worker:

https://github.com/DataBiosphere/toil/blob/0446fe285fe57a3d920d4b5a30d00f2a5aa8589c/src/toil/worker.py#L192-L228

Now, Slurm's `sbatch` will additionally do its own environment capture and restore, and single-machine Toil also will end up inheriting the leader's environment for free anyway, so it's possible this system is somehow broken and we haven't noticed.

**Comment by glennhickey** [comment]
Thanks @adamnovak .  This does seem to be a Toil bug then.  It reproduces quite quickly with
```
virtualenv --system-site-packes cactus-venv
. cactus-venv/bin/activate
wget https://github.com/ComparativeGenomicsToolkit/cactus/releases/download/v2.7.2/cactus-bin-v2.7.2.tar.gz
tar zxf cactus-bin-v2.7.2.tar.gz
cd cactus-bin-v2.7.2
pip install -U .
cactus aws:us-west-2:glennhickey-jobstore11 ./examples/evolverMammals.txt em.hal --provisioner aws --batchSystem mesos --nodeTypes c4.8xlarge --skipPreprocessor --consCores 4
```
which fails with the same error reported above.  It happens because it can't find the `CACTUS_MAX_MEMORY` and `CACTUS_DEFAULT_MEMORY` environment variables which cactus sets itself before `start()`.  I also tried to export them on the command line first and got the same result.

**Comment by glennhickey** [comment]
@tinaveit This is a bug in cactus, not Toil as @adamnovak [pointed out](https://github.com/DataBiosphere/toil/issues/4835#issuecomment-1992554248).  

The good news is that I think you should be able to resume your workflow by running with `--restart --defaultMemory 2000000000 --maxMemory 256000000000`.  Thanks for raising this, I'll make sure it's fixed in the next Cactus release (which should be pretty soon).

**Comment by adamnovak** [comment]
@tinaveit I think we have sorted out all the problems reported here, so I'm going to close this. If you hit another problem in the same workflow, can you open a new issue and reference this one?
